Phuc

EsercitoHaki opened the pull request:
Add Lesson management features including DTO, Repository, Service

Commit:
Add Lesson management features including DTO, Repository, Service, and Controller

Commit:
feat: Implement Question Management System

- Added Question model with relationships to Lesson.
- Created QuestionRepository and QuestionRepositoryInterface for data access.
- Developed QuestionService and QuestionServiceInterface for business logic.
- Implemented QuestionController for handling API requests related to questions.
- Created migrations for questions table and removed unnecessary correct_answer column.
- Added seeders for populating sample questions and lessons.
- Defined routes for question-related API endpoints.
- Introduced QuestionDTO for data transfer between layers.

Commit:
feat: Implement options and user progress management

- Added CreateOptionsRequest and OptionRequest for validation of option creation and updates.
- Created Option and UserProgress models with relationships to questions and users respectively.
- Developed OptionRepository and UserProgressRepository for data handling and business logic.
- Introduced services (OptionService and UserProgressService) to encapsulate business logic for options and user progress.
- Implemented seeder for populating options data in the database.
- Defined API routes for managing options and user progress, including CRUD operations and statistics retrieval.

Commit:
feat: Show explanation for correct answers and submit each answer individually instead of merging into one API, including explanation display for correct answers

Commit:
feat: Implement friend management system with friend requests and friendships

EsercitoHaki commented on app/Http/Controllers/Api/V1/FriendController.php:
```diff
@@ -0,0 +1,163 @@
+        if ($existingRequest) {
+            return response()->json([
+                'success' => false,
+                'message' => 'You already have a pending friend request to this user.'
```
Đã chốt là message là tiếng việt còn gì ?

EsercitoHaki commented on app/Models/Friend.php:
```diff
@@ -0,0 +1,73 @@
+     * The table associated with the model.
+     *
+     * @var string
+     */
```
Comment thì tiếng việt PHPDoc comment thì tiếng anh ?, để 1 kiểu thôi (tiếng việt), không thì xóa đi

EsercitoHaki commented on app/Services/FriendService.php:
```diff
@@ -0,0 +1,269 @@
+                'message' => 'Lời mời kết bạn không tồn tại'
+            ];
+        }
+
```
Lặp lại quá nhiều các kiểm tra tồn tại của yêu cầu kết bạn, quyền hạn,.. , nên tạo 1 phương thức dùng chung để gọi lại, thay vì cứ lặp đi lặp lại

EsercitoHaki commented on app/Http/Controllers/Api/V1/FriendController.php:
```diff
@@ -0,0 +1,163 @@
+        $result = $this->friendService->acceptFriendRequest($requestId);
+
+        return response()->json($result, $result['success'] ? 200 : 400);
+    }
```
Trả về dữ liệu phải nhất quán chứ, trong các phương thức khác ông trả về JSON (message, data, sucsess) ở đây là trả về thẳng từ Service, ở một số phương thức khác thì trả về JSON nhưng chỗ thì thiếu data, chỗ thì thiếu message ?

EsercitoHaki commented on app/Http/Requests/CreateOptionsRequest.php:
```diff
@@ -0,0 +1,30 @@
+            'options.*.is_correct' => ['required', 'boolean']
+        ];
+    }
+}
```
Nên thêm các validation message cho những exception liên quan tới data, chứ đừng lạm dụng exception

EsercitoHaki commented on app/Models/Friend.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        return $this->belongsTo(User::class, 'user2_id', 'user_id');
+    }
+}
```
Model thì khá ổn rồi

EsercitoHaki reviewed (commented):
Chỉnh sửa lại phần message cũng như commet cho thống nhất (tiếng việt), thêm 1 số phương thức dùng chung để tránh lặp lại code, một số chỗ cần thêm try catch tránh lỗi 500, cải thiện các truy vấn

anvinh2812 commented:
Không chạy được luồng làm bài (lỗi 404)
![image](https://github.com/user-attachments/assets/a31ab259-a31c-4f4a-967c-5d3a8b27f772)

Commit:
refactor: chuyển đổi từ exception sang validation message cho data
- Thêm các phương thức validation trước khi thực hiện thao tác thay vì phụ thuộc vào exception
- Cải thiện xử lý lỗi trong FriendService với thông báo lỗi tùy chỉnh
- Thêm các phương thức hỗ trợ kiểm tra dữ liệu trong repositories
- Cập nhật interface để khớp với các phương thức mới trong lớp thực thi
- Tối ưu hiệu suất bằng cách giảm việc sử dụng exception cho xác thực dữ liệu

Commit:
feat: Enhance Lesson management with improved error handling, validation, and new methods for category management

Commit:
refactor: UserProgress functionality and improve error handling

- Enhanced UserProgressController to include try-catch blocks for better error handling and logging.
- Updated response structure to provide consistent success and error messages.
- Added validation for user answers in completeLesson and submitAnswer methods.
- Introduced new methods in UserProgress model to check lesson progress and completion status.
- Improved UserProgressRepository with error handling and logging.
- Added methods to retrieve completed and in-progress lessons.
- Enhanced UserProgressService with error handling and validation for user inputs.
- Updated UserProgressServiceInterface to include new methods for checking lesson status.

Commit:
refactor: Tái cấu trúc interfaces vào thư mục riêng biệt
- Di chuyển service interfaces vào thư mục App\Services\Interfaces
- Di chuyển repository interfaces vào thư mục App\Repositories\Interfaces
- Cập nhật namespaces và imports trong các controllers, services và repositories
- Cập nhật bindings trong RepositoryServiceProvider với namespace mới
- Sửa lỗi  500  "error": "Call to undefined relationship [options] on model [App\\Models\\Question].""
